rir/src/interpreter/interp.cpp

```diff
@@ -2510,6 +2525,81 @@ SEXP evalRirCode(Code* c, InterpreterInstance* ctx, SEXP env,
             NEXT();
         }
 
+        INSTRUCTION(asint_) {
+            bool ceil_ = (bool)readImmediate();
+            advanceImmediate();
+            SEXP val = ostack_top(ctx);
+            // Scalar integers (already done)
+            if (IS_SIMPLE_SCALAR(val, INTSXP)) {
+                NEXT();
+            }
+            int x = -20;
+            bool isNaOrNan = false;
+            if (TYPEOF(val) == INTSXP || TYPEOF(val) == REALSXP ||
+                TYPEOF(val) == LGLSXP) {
+                if (XLENGTH(val) == 0) {
+                    Rf_errorcall(getSrcAt(c, pc - 1, ctx),
+                                 "argument of length 0");
+                    x = NA_INTEGER;
+                    isNaOrNan = false;
+                } else {
+                    switch (TYPEOF(val)) {
+                    case INTSXP: {
+                        int i = *INTEGER(val);
+                        if (i == NA_INTEGER) {
+                            x = NA_INTEGER;
+                            isNaOrNan = true;
+                        } else {
+                            x = i;
+                            isNaOrNan = false;
+                        }
+                        break;
+                    }
+                    case REALSXP: {
+                        double r = *REAL(val);
+                        if (ISNAN(r)) {
+                            x = NA_INTEGER;
+                            isNaOrNan = true;
+                        } else {
+                            x = (int)(ceil_ ? ceil(r) : floor(r));
+                            isNaOrNan = false;
+                        }
+                        break;
+                    }
+                    case LGLSXP: {
+                        int l = *LOGICAL(val);
+                        if (l == NA_LOGICAL) {
+                            x = NA_INTEGER;
+                            isNaOrNan = true;
+                        } else {
+                            x = (int)l;
+                            isNaOrNan = false;
+                        }
+                        break;
+                    }
+                    default:
+                        assert(false);
+                    }
+                    if (XLENGTH(val) > 1) {
+                        Rf_warningcall(getSrcAt(c, pc - 1, ctx),
+                                       "numerical expression has multiple "
+                                       "elements: only the first used");
+                    }
+                }
+            } else { // Everything else
+                x = NA_INTEGER;
+                isNaOrNan = true;
+            }
+            if (isNaOrNan) {
+                Rf_errorcall(getSrcAt(c, pc - 1, ctx), "NA/NaN argument");
+            }
+            res = Rf_allocVector(INTSXP, 1);
```

Comment:
here you could check, if `NO_REFERENCES(val)` and also val is a simple scalar, you can store the result into val instead of creating a new vector.